Mathlib/LinearAlgebra/LinearDisjoint.lean

```diff
@@ -0,0 +1,550 @@
+    {a b : S} (ha : a ∈ M ⊓ N) (hb : b ∈ M ⊓ N) (hc : Commute a b) :
+    ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat' (Or.inl ‹_›) ha hb hc
+
+theorem not_linearIndependent_pair_of_commute_of_flat_right' [Module.Flat R N]
+    {a b : S} (ha : a ∈ M ⊓ N) (hb : b ∈ M ⊓ N) (hc : Commute a b) :
+    ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat' (Or.inr ‹_›) ha hb hc
+
+end
+
+theorem rank_inf_le_one_of_commute_of_flat (hf : Module.Flat R M ∨ Module.Flat R N)
+    (hc : ∀ (m n : ↥(M ⊓ N)), Commute m.1 n.1) : Module.rank R ↥(M ⊓ N) ≤ 1 := by
+  nontriviality R
+  refine rank_le fun s h ↦ ?_
+  by_contra hs
+  rw [not_le, ← Fintype.card_coe, Fintype.one_lt_card_iff_nontrivial] at hs
+  obtain ⟨a, b, hab⟩ := hs.exists_pair_ne
+  refine H.not_linearIndependent_pair_of_commute_of_flat hf a.1 b.1 (hc a.1 b.1) ?_
+  have := h.comp ![a, b] fun i j hij ↦ by
+    fin_cases i <;> fin_cases j
+    · rfl
+    · simp [hab] at hij
+    · simp [hab.symm] at hij
+    · rfl
+  convert this
+  ext i
+  fin_cases i <;> simp
+
+theorem rank_inf_le_one_of_commute_of_flat_left [Module.Flat R M]
+    (hc : ∀ (m n : ↥(M ⊓ N)), Commute m.1 n.1) : Module.rank R ↥(M ⊓ N) ≤ 1 :=
+  H.rank_inf_le_one_of_commute_of_flat (Or.inl ‹_›) hc
+
+theorem rank_inf_le_one_of_commute_of_flat_right [Module.Flat R N]
+    (hc : ∀ (m n : ↥(M ⊓ N)), Commute m.1 n.1) : Module.rank R ↥(M ⊓ N) ≤ 1 :=
+  H.rank_inf_le_one_of_commute_of_flat (Or.inr ‹_›) hc
+
+theorem rank_le_one_of_commute_of_flat_of_self (H : M.LinearDisjoint M) [Module.Flat R M]
+    (hc : ∀ (m n : M), Commute m.1 n.1) : Module.rank R M ≤ 1 := by
+  rw [← inf_of_le_left (le_refl M)] at hc ⊢
+  exact H.rank_inf_le_one_of_commute_of_flat_left hc
+
+end not_linearIndependent_pair
+
+end LinearDisjoint
+
+end Ring
+
+section CommRing
+
+namespace LinearDisjoint
+
+variable [CommRing R] [CommRing S] [Algebra R S]
+
+variable (M N : Submodule R S)
+
+section not_linearIndependent_pair
+
+variable {M N}
+
+variable (H : M.LinearDisjoint N)
+
+section
+
+variable [Nontrivial R]
+
+theorem not_linearIndependent_pair_of_flat_left [Module.Flat R M]
+    (a b : ↥(M ⊓ N)) : ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat_left a b (mul_comm _ _)
+
+theorem not_linearIndependent_pair_of_flat_right [Module.Flat R N]
+    (a b : ↥(M ⊓ N)) : ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat_right a b (mul_comm _ _)
+
+theorem not_linearIndependent_pair_of_flat (hf : Module.Flat R M ∨ Module.Flat R N)
+    (a b : ↥(M ⊓ N)) : ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat hf a b (mul_comm _ _)
+
+theorem not_linearIndependent_pair_of_flat' (hf : Module.Flat R M ∨ Module.Flat R N)
+    {a b : S} (ha : a ∈ M ⊓ N) (hb : b ∈ M ⊓ N) : ¬LinearIndependent R ![a, b] :=
+  H.not_linearIndependent_pair_of_commute_of_flat' hf ha hb (mul_comm _ _)
```

Comment:
Do we need to specialize these? I thought they were helper lemmas.

Reply:
Do you mean `flat` version vs `flat'` version? The `flat` is bundled version (i.e. `a b : ↥(M ⊓ N)`) while the `flat'` version is not (i.e. `a b : S` with the condition that they are `∈ M ⊓ N`). But maybe you are right, I checked the downstream code and it looks like that the `flat'` version is unused. Do you think it's a good idea to remove `flat'` version?

Reply:
`flat'` version removed.